tests/e2e/specs/backend/product-query/atomic-blocks.test.ts

```diff
@@ -0,0 +1,127 @@
+/**
+ * External dependencies
+ */
+import { canvas, setPostContent, insertBlock } from '@wordpress/e2e-test-utils';
+import { visitBlockPage, saveOrPublish } from '@woocommerce/blocks-test-utils';
+
+/**
+ * Internal dependencies
+ */
+import { GUTENBERG_EDITOR_CONTEXT, describeOrSkip } from '../../../utils';
+import { shopper } from '../../../../utils';
+
+const block = {
+	name: 'Product Query',
+	slug: 'woocommerce/product-query',
+	class: '.wp-block-query',
+};
+
+describeOrSkip( GUTENBERG_EDITOR_CONTEXT === 'gutenberg' )(
+	'Product Query > Atomic blocks',
+	() => {
+		beforeEach( async () => {
+			await visitBlockPage( `${ block.name } Block` );
+		} );
+		afterAll( async () => {
+			await visitBlockPage( `${ block.name } Block` );
+			await setPostContent( '' );
+			await insertBlock( 'Product Query' );
+			await saveOrPublish();
+		} );
+
+		it( 'Can add the Add to Cart Button block and render it on the front end.', async () => {
+			const canvasEl = canvas();
+			await canvasEl.waitForSelector(
+				`${ block.class } .wp-block-woocommerce-product-image a img`
+			);
+			const postTitleEl = await canvasEl.$(
+				`${ block.class } .wp-block-post-title`
+			);
+			await postTitleEl.click();
+			await insertBlock( 'Add to Cart Button' );
+			await expect( canvasEl ).toMatchElement( block.class, {
+				text: 'Add to cart',
+			} );
+			await saveOrPublish();
+
+			await shopper.block.goToBlockPage( block.name );
+			await page.waitForSelector(
+				'.wp-block-query button.add_to_cart_button'
+			);
+			await expect( page ).toClick( 'button', {
+				text: 'Add to cart',
+			} );
+			await shopper.block.goToCart();
+			await page.waitForSelector( '.wc-block-cart-items__row' );
+			expect(
+				await page.$$eval(
+					'.wc-block-cart-items__row',
+					( rows ) => rows.length
+				)
+			).toEqual( 1 );
+		} );
+
+		it( 'Can add the Product Image block', async () => {
+			const canvasEl = canvas();
+			await canvasEl.waitForSelector(
+				`${ block.class } .wp-block-woocommerce-product-image a img`
+			);
+			const postTitleEl = await canvasEl.$(
+				`${ block.class } .wp-block-post-title`
+			);
+			await postTitleEl.click();
+			await insertBlock( 'Product Image' );
+			expect(
+				await canvasEl.$$eval(
+					`${ block.class } ul.wp-block-post-template > li.block-editor-block-list__layout .wp-block-woocommerce-product-image`,
+					( images ) => images.length
+				)
+			).toEqual( 2 );
+		} );
+
+		it( 'Can add the Product Price block and render it on the front end', async () => {
+			const canvasEl = canvas();
+			await canvasEl.waitForSelector(
+				`${ block.class } .wp-block-woocommerce-product-image a img`
+			);
+			const postTitleEl = await canvasEl.$(
+				`${ block.class } .wp-block-post-title`
+			);
+			await postTitleEl.click();
```

Comment:
It seems that you are relying on clicking the post title to insert the block in the right place. We might migrate away from using the Post Title (and use “Product title”, for example), or certain patterns might not have that title at all. So, it seems a bit of a brittle way to describe this test, is there any other way to insert an inner block more confidently?

If not, at least the refactoring of this implementation details in a function as described in my other comment will make it easier to change it in the future. But I'd like to see if we can find a safer way to do this.

Reply:
I agree, I created `insertInnerBlock` as a safer way to insert inner block. It work with all parent blocks.